File: cpp/src/detection/detection.cpp
#include <iostream>
#include <fstream>
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

class ObjectDetector
{
public:
    ObjectDetector(const std::string& directoryName, const std::string& sourcePath);

    int runObjectDetection();

private:
    // program input parameters
    std::string directoryName_;
    std::string sourcePath_;

    // path to files
    std::string path_video_input;
    std::string path_video_output;
    std::string path_class_input;
    std::string path_net_input;
    // constants
    const std::vector<cv::Scalar> colors = {cv::Scalar(255, 255, 0), cv::Scalar(0, 255, 0), cv::Scalar(0, 255, 255), cv::Scalar(255, 0, 0)};

    const float INPUT_WIDTH = 640.0;
    const float INPUT_HEIGHT = 640.0;
    const float SCORE_THRESHOLD = 0.2;
    const float NMS_THRESHOLD = 0.45;
    const float CONFIDENCE_THRESHOLD = 0.25;

    // structs
    struct Detection
    {
        int class_id;
        float confidence;
        cv::Rect box;
    };

    // video input details
    double input_fps;
    int fw;
    int fh;

    // variables
    std::vector<std::string> class_list;
    int total_frames = 0;

    // methods
    void load_class_list(std::vector<std::string>& class_list);

    void load_net(cv::dnn::Net& net);

    cv::Mat format_yolov5(const cv::Mat& source);

    void detect(cv::Mat& image, cv::dnn::Net& net, std::vector<Detection>& output,
        const std::vector<std::string>& className);

    void drawBBox(cv::Mat& frame, std::vector<Detection>& output, const std::vector<std::string>& class_list);

    // init capture, writer and network
    cv::VideoCapture cap;
    cv::VideoWriter out;
    cv::dnn::Net net;
};

void ObjectDetector::load_class_list(std::vector<std::string> &class_list)
{
    std::ifstream ifs(path_class_input);
    std::string line;
    while (getline(ifs, line))
    {
        class_list.push_back(line);
    }
}

void ObjectDetector::load_net(cv::dnn::Net &net)
{   
    cv::dnn::Net result = cv::dnn::readNetFromONNX(path_net_input);
    if (cv::cuda::getCudaEnabledDeviceCount())
    {
        std::cout << "Running with CUDA\n";
        result.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
        result.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA_FP16);
    }
    else
    {
        std::cout << "Running on CPU\n";
        result.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
        result.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
    }
    net = result;
}

ObjectDetector::ObjectDetector(const std::string& directoryName, const std::string& sourcePath)
    : directoryName_(directoryName), sourcePath_(sourcePath)
{
    // Concatenate the directory name with another string
    path_video_input = sourcePath;
    path_video_output = directoryName + "output/detect_cpp.mp4";
    path_class_input = directoryName + "classes/classes_train.txt";
    path_net_input = directoryName + "models/best_all.onnx";

    // Open video input
    cap.open(path_video_input);

    if (!cap.isOpened())
    {
        std::cerr << "Error opening video file\n";
    }

    // Prints video input info
    input_fps = cap.get(cv::CAP_PROP_FPS);
    fw = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
    fh = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
    std::cout << "FPS: " << input_fps << ", Width: " << fw << ", Height: " << fh << std::endl;

    // Open video output
    out.open(path_video_output, cv::VideoWriter::fourcc('m', 'p', '4', 'v'), input_fps, cv::Size(fw, fh));

    if (!out.isOpened())
    {
        std::cerr << "Error creating VideoWriter\n";
    }

    // Load class list
    load_class_list(class_list);

    // Load net
    load_net(net);
}

// cv::Mat ObjectDetector::format_yolov5(const cv::Mat& source) {
//     int col = source.cols;
//     int row = source.rows;
//     int _max = MAX(col, row);
//     cv::Mat result = cv::Mat::zeros(_max, _max, CV_8UC3);
//     source.copyTo(result(cv::Rect(0, 0, col, row)));
//     return result;
// }

void ObjectDetector::detect(cv::Mat& image, cv::dnn::Net& net, std::vector<Detection>& output, const std::vector<std::string>& className) {
    cv::Mat blob;

    // auto input_image = format_yolov5(image);
    cv::dnn::blobFromImage(image, blob, 1./255., cv::Size(INPUT_WIDTH, INPUT_HEIGHT), cv::Scalar(), true, false);

    // forward pass into network
    net.setInput(blob);
    std::vector<cv::Mat> outputs;
    net.forward(outputs, net.getUnconnectedOutLayersNames());

    float x_factor = image.cols / INPUT_WIDTH;
    float y_factor = image.rows / INPUT_HEIGHT;
    
    float *data = (float *)outputs[0].data;

    const int dimensions = 85;
    const int rows = 25200;
    
    std::vector<int> class_ids;
    std::vector<float> confidences;
    std::vector<cv::Rect> boxes;

    // unwrap detections
    for (int i = 0; i < rows; ++i) {
        float confidence = data[4]; // conf in 4th address
        if (confidence >= CONFIDENCE_THRESHOLD) {
            float * classes_scores = data + 5; // address of class scores start 5 addresses away
            cv::Mat scores(1, className.size(), CV_32FC1, classes_scores); // create mat for score per detection
            cv::Point class_id;
            double max_class_score;
            minMaxLoc(scores, 0, &max_class_score, 0, &class_id); // find max score
            // if (max_class_score > SCORE_THRESHOLD) {

                confidences.push_back(confidence); // add conf to vector
                class_ids.push_back(class_id.x); // add class_id to vector
                
                // center coords of bbox
                float x = data[0];
                float y = data[1];
                float w = data[2];
                float h = data[3];
                int left = int((x - 0.5 * w) * x_factor);
                int top = int((y - 0.5 * h) * y_factor);
                int width = int(w * x_factor);
                int height = int(h * y_factor);
                boxes.push_back(cv::Rect(left, top, width, height));
            // }
        }
        data += 85; // next detection (x,y,w,h,conf,80 class conf)
    }
    // nms
    std::vector<int> nms_result;
    cv::dnn::NMSBoxes(boxes, confidences, SCORE_THRESHOLD, NMS_THRESHOLD, nms_result);
    for (int i = 0; i < nms_result.size(); i++) {
        int idx = nms_result[i];
        Detection result;
        result.class_id = class_ids[idx];
        result.confidence = confidences[idx];
        result.box = boxes[idx];
        output.push_back(result); // add to output
    }
}

void ObjectDetector::drawBBox(cv::Mat& frame, std::vector<Detection>& output, const std::vector<std::string>& class_list)
{
    int detections = output.size();

    for (int i = 0; i < detections; ++i)
    {
        auto detection = output[i];
        auto box = detection.box;
        auto classId = detection.class_id;
        const auto color = colors[classId % colors.size()];
        cv::rectangle(frame, box, color, 3);
        cv::rectangle(frame, cv::Point(box.x, box.y - 20), cv::Point(box.x + box.width, box.y), color, cv::FILLED);
        cv::putText(frame, class_list[classId].c_str() + std::to_string(detection.confidence), cv::Point(box.x, box.y - 5), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 0));
    }
}
int ObjectDetector::runObjectDetection()
{ 
    cv::Mat frame;

    while (true)
    {
        auto start = std::chrono::high_resolution_clock::now();

        cap.read(frame);

        if (frame.empty())
        {
            std::cout << "End of stream\n";
            break;
        }

        std::vector<Detection> output;
        detect(frame, net, output, class_list);
        drawBBox(frame, output, class_list);
        total_frames++;
        cv::imshow("output", frame);
        out.write(frame);

        if (cv::waitKey(1) != -1)
        {
            cap.release();
            std::cout << "finished by user\n";
            break;
        }

        auto end = std::chrono::high_resolution_clock::now();
        auto time_taken = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
        std::cout << "Time taken for detection in ms: " << time_taken  << std::endl;
    }

    std::cout << "Total frames: " << total_frames << "\n";
    cap.release();
    out.release();
    cv::destroyAllWindows();
    return 0;
}

int main(int argc, char *argv[])
{
    // Check if all three arguments are provided
    if (argc < 3)
    {
        std::cout << "Please provide /path/to/tracking_ws/ and /path/to/source" << std::endl;
        return 1;
    }

    // Get the directory name, source path, and tracker name from the arguments
    std::string directoryName = argv[1];
    std::string sourcePath = argv[2];

    // Concatenate the directory name with another string
    std::string concatenatedString = directoryName + "/another_string";

    ObjectDetector detector(directoryName, sourcePath);

    detector.runObjectDetection();

    return 0;
}

File: python/src/detection.py
import cv2 
import numpy as np
from helper_funcs import *
import argparse
import os

# Create the argument parser
parser = argparse.ArgumentParser(description='Detection engine written in python with opencv')

# Add the input and output file arguments
parser.add_argument('tracking_ws_path', help='/path/to/tracking_ws/')
parser.add_argument('input_video_path', help='/path/to/video/file/video.xxx')


# Parse the arguments
args = parser.parse_args()
input_video_path = args.input_video_path
tracking_ws_path = args.tracking_ws_path
video_name = input_video_path.split('/')[-1]
video_name_no_ext = video_name.split('.')[0]
output_video_path = os.path.join(tracking_ws_path,'output',video_name_no_ext+'_detect_py.mp4')
print(output_video_path)

class objectDetector():

    def __init__(self):
        # Constan-ts
        self.detect_conf_thres = 0.4
        self.class_conf_thres = 0.25
        self.nms_thres = 0.4

        # Variables
        self.frame_count = 0

        # Video I/O
        self.cap = cv2.VideoCapture(input_video_path) # Create a VideoCapture object
        # self.cap = cv2.VideoCapture(0)  
        self.fps = self.cap.get(cv2.CAP_PROP_FPS)
        self.fw = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        self.fh = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        print(f"FPS: {self.fps}, Width: {self.fw}, Height: {self.fh}")
        self.out = cv2.VideoWriter(output_video_path,cv2.VideoWriter_fourcc('m','p','4','v'),self.fps,(self.fw,self.fh)) # create writer obj

        # Detector init
        self.net = cv2.dnn.readNet('/home/jeric/tracking_ws/models/last_exp2.onnx') # input obj detector network

        self.net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
        self.net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA)

        # Data storage
        self.detect_bboxes = []
        self.detect_conf = []
        self.detect_class_ids = []


    def detect(self, current_frame):
        print('Detecting Objects...')

        # format image in to fit 640x640 and colour
        blob, input_image = format_yolov5(current_frame) 

        # set formatted image into net and pass through net to obtain output
        self.net.setInput(blob)
        predictions = self.net.forward()
        output = predictions[0]

        # unwrap detections into usable format
        bboxes, confidences, class_ids = unwrap_detection_numpy(input_image,output, self.detect_conf_thres, self.class_conf_thres)

        # NMS to remove dup and overlap
        result_bboxes, result_confidences, result_class_ids = nms(bboxes, confidences, class_ids, self.detect_conf_thres, self.nms_thres)

        self.detect_bboxes = result_bboxes
        self.detect_conf = result_confidences
        self.detect_class_ids = result_class_ids
    

def main(args=None):
    object_detector = objectDetector()
    cv2.namedWindow("camera", cv2.WINDOW_NORMAL)
    cv2.resizeWindow("camera", 1280, 720)

    while 1:
        # Start the timer
        start_time = cv2.getTickCount()

        ret, current_frame = object_detector.cap.read()

        if not ret:
            break
             
        # pass through detector
        object_detector.detect(current_frame)
        draw_bbox(current_frame, object_detector.detect_bboxes, object_detector.detect_class_ids, tracking=False)

        # Write the frame into the file 'output.avi' 
        object_detector.out.write(current_frame)
        
        cv2.imshow("camera", current_frame) # Display image
        cv2.waitKey(1)
        
        # Calculate the elapsed time
        ticks = cv2.getTickCount() - start_time
        elapsed_time = (ticks / cv2.getTickFrequency()) * 1000  # Convert to milliseconds
        print("Time taken for Detection:", elapsed_time, "milliseconds\n")


if __name__ == '__main__':
    main()
